src/storage.rs

```diff
@@ -0,0 +1,151 @@
+pub struct AddressOffset<U>(pub U);
+
+use core::ops::{Add, Sub};
+
+/// Implement add for the Address and AddressOffset Types.
+impl<'a, 'b, U> Add<&'b AddressOffset<U>> for &'a Address<U>
+where
+    U: Add<U, Output = U> + Copy,
+{
+    type Output = Address<U>;
+
+    fn add(self, other: &'b AddressOffset<U>) -> Address<U> {
+        Address(self.0 + other.0)
+    }
+}
+
+/// Implement subtraction for the Address and AddressOffset Types.
+impl<'a, 'b, U> Sub<&'b AddressOffset<U>> for &'a Address<U>
+where
+    U: Sub<U, Output = U> + Copy,
+{
+    type Output = Address<U>;
+
+    fn sub(self, other: &'b AddressOffset<U>) -> Address<U> {
+        Address(self.0 - other.0)
+    }
+}
+
+/// Page represents an unsigned integer that is a Page ID in the device memory space.
+pub struct Page<U>(pub U);
+
+/// Read a single word from the device.
+///
+/// `Word` type allows any word size to be used.
+pub trait SingleRead<Word, U> {
+    /// An enumeration of Storage errors
+    type Error;
+
+    /// Reads the word stored at the address
+    /// ```
+    /// pub fn try_read(&mut self, address: Address) -> nb::Result<u8, Self::Error>
+    ///     let address = address.0 as *const _;
+    ///     unsafe {
+    ///         Ok(core::slice::from_raw_parts::<'static, u8>(address,length))
+    ///     }
+    /// ```
+    fn try_read(&mut self, address: Address<U>) -> nb::Result<Word, Self::Error>;
+}
+
+/// Write a single word to the device.
+///
+/// `Word` type allows any word size to be used.
+pub trait SingleWrite<Word, U> {
+    /// An enumeration of Storage errors
+    type Error;
+
+    /// Writes the word to the address
+    fn try_write(&mut self, address: Address<U>, word: Word) -> nb::Result<(), Self::Error>;
+}
+
+/// Read multiple bytes from the device.
+///
+/// intended to be used for when there is a optimized method of reading multiple bytes.
+///
+/// Iterating over the slice is a valid method to ```impl``` this trait.
+pub trait MultiRead<Word, U> {
+    /// An enumeration of Storage errors
+    type Error;
+
+    /// Reads the words stored at the address to fill the buffer
+    /// ```
+    /// pub fn try_read_slice(
+    ///     &mut self,
+    ///     address: Address,  
+    ///     buf: &mut [Word]
+    /// ) -> nb::Result<Word, Self::Error>
+    ///     let address = address.0 as *const _;
+    ///     unsafe {
+    ///         buf = core::slice::from_raw_parts::<'static, Word>(address,buf.len())
+    ///     }
+    ///     
+    ///     Ok()
+    /// }
+    /// ```
+    fn try_read_slice(
+        &mut self,
+        address: Address<U>,
+        buf: &mut [Word],
+    ) -> nb::Result<(), Self::Error>;
+}
+
+/// Write multiple bytes to the device.
+///
+/// intended to be used for when there is a optimized method of reading multiple bytes.
+///
+/// Iterating over the slice is a valid method to ```impl``` this trait.
+pub trait MultiWrite<Word, U> {
```

Comment:
What is the expected behavior of write? ANDed like flash, always as is or device specific?